models/post.go

```diff
@@ -143,36 +152,74 @@ func (p *Post) getPostTags() error {
 	return nil
 }
 
-// GetPostsByScoreSince - get posts from the database that have been created since
-// the provided cutoff time, and offset the results
-func GetPostsByScoreSince(cutoff time.Time, offset int) ([]Post, error) {
-	posts := []Post{}
+// GetPostTags - get the tags with post id
+func GetPostTags(id int) ([]PostTag, error) {
+	tags := []PostTag{}
 
-	err := DBConn.Select(&posts, "SELECT * FROM `posts` WHERE created_at > ? ORDER BY `score` DESC LIMIT 100 OFFSET ?", cutoff.Format("2006-01-02 15:04:05"), offset)
+	err := DBConn.Select(&tags, "SELECT * FROM posts_tags where post_id = ?", id)
+	if err != nil {
+		return nil, err
+	}
 
-	return posts, err
+	// query the tag name with id
+	for i, item := range tags {
+		tag := Tag{}
+		if err = tag.FindByID(item.TagID); err != nil {
+			return nil, err
+		}
+		tags[i].TagName = tag.Name
+	}
+
+	return tags, err
 }
 
-// GetPostsByPostTagScoreSince - get the popular posts from the database that have been create since
-// 24 hours before, and tag name, and score order, and time cutoff, limit 100
-func GetPostsByPostTagScoreSince(name string, cutoff time.Time) ([]Post, error) {
-	posts := []Post{}
+func intSlice2Str(vals []int) string {
+	var s []string
 
-	since := cutoff.Format("2006-01-02 15:04:05")
+	for _, v := range vals {
+		s = append(s, strconv.Itoa(v))
+	}
+	return strings.TrimSuffix(strings.Join(s, ","), ",")
+}
 
-	err := DBConn.Select(&posts, "select * from posts where created_at > ? and id in (select t1.post_id from posts_tags t1 join tags t2 on t1.tag_id = t2.id and t2.name = ? order by t1.score desc) limit 100;", since, name)
+// GetPostsByParams - get the posts by parameters
+func GetPostsByParams(params *PostQueryParams) ([]*Post, error) {
+	args := []interface{}{}
 
-	return posts, err
-}
+	query := "select * from posts where created_at > ?"
+	// time range
+	args = append(args, params.Since)
+
+	// special user
+	if params.UserID > 0 {
+		query = query + " and user_id = ?"
+		args = append(args, params.UserID)
+	}
+
+	// tags
+	if len(params.PostIDs) > 0 {
+		query = query + " and id in (?)"
+		args = append(args, intSlice2Str(params.PostIDs))
+	}
+
+	// orders
+	if params.SortedByScore {
+		query = query + " order by score desc"
```

Comment:
One exception to this is if the query is filtered to a tag, then the sort will be by the score of the posttag, not the score of the post. You'll likely have to set a flag in the previous section and look for it here.